nexus/db-queries/src/db/datastore/inventory.rs

```diff
@@ -0,0 +1,1568 @@
+                                        found_caboose.caboose.version.clone(),
+                                    )),
+                            ),
+                        )
+                        .select((
+                            dsl_baseboard_id::id,
+                            dsl_sw_caboose::id,
+                            collection_id.into_sql::<diesel::sql_types::Uuid>(),
+                            found_caboose
+                                .time_collected
+                                .into_sql::<diesel::sql_types::Timestamptz>(),
+                            found_caboose
+                                .source
+                                .clone()
+                                .into_sql::<diesel::sql_types::Text>(),
+                            db_which.into_sql::<CabooseWhichEnum>(),
+                        ));
+
+                    let _ = diesel::insert_into(db::schema::inv_caboose::table)
+                        .values(selection)
+                        .into_columns((
+                            dsl_inv_caboose::hw_baseboard_id,
+                            dsl_inv_caboose::sw_caboose_id,
+                            dsl_inv_caboose::inv_collection_id,
+                            dsl_inv_caboose::time_collected,
+                            dsl_inv_caboose::source,
+                            dsl_inv_caboose::which,
+                        ))
+                        .execute_async(&conn)
+                        .await?;
+
+                    // See the comments above.  The same applies here.  If you
+                    // update the statement below because the schema for
+                    // `inv_caboose` has changed, be sure to update the code
+                    // above, too!
+                    let (
+                        _hw_baseboard_id,
+                        _sw_caboose_id,
+                        _inv_collection_id,
+                        _time_collected,
+                        _source,
+                        _which,
+                    ) = dsl_inv_caboose::inv_caboose::all_columns();
+                }
+            }
+
+            // Finally, insert the list of errors.
+            {
+                use db::schema::inv_collection_error::dsl as errors_dsl;
+                let _ = diesel::insert_into(errors_dsl::inv_collection_error)
+                    .values(error_values)
+                    .execute_async(&conn)
+                    .await?;
+            }
+
+            Ok(())
+        })
+        .await
+        .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        info!(
+            &opctx.log,
+            "inserted inventory collection";
+            "collection_id" => collection.id.to_string(),
+        );
+
+        Ok(())
+    }
+
+    /// Prune inventory collections stored in the database, keeping at least
+    /// `nkeep`.
+    ///
+    /// This function removes as many collections as possible while preserving
```

Comment:
The latest `nkeep` are provided by timestamps, which aren't really global. Right now collection is at 10 min intervals, so as long as only 1 nexus performs a collection per interval this should be fine. However, I could see problems arising around order, although quite unlikely due to our 500ms limitation around syncing. 

I don't really think this is something worth worrying about but figured I'd ask for completeness sake. Are there autoincrementing IDs we could use for collections rather than UUIDs as foreign keys and then sort by those? Would this present other issues with dueling Nexuses?

Reply:
Yeah, using timestamps is definitely a little fuzzy.  In this case though I think that reflects the reality that collections are not atomic and they don't have a total order.  Two Nexus instances could totally run collections concurrently that have start/done times that overlap (and I think that's fine).  Consumers can decide if they want the most-recently-started or most-recently-finished (or even the-one-containing-the-most-recent-collection-time-for-the-specific-item-that-I-care-about).  We could potentially use a [sequence](https://www.cockroachlabs.com/docs/stable/create-sequence) to assign a total order to these but I don't think it would have a useful semantic meaning -- at best it'd be a proxy for "which one committed to the database first" and I'm not sure that's useful.

Okay so my argument is basically "report the facts (the start/done timestamp) and let consumers decide what they want".  But that just punts your question to "okay, well, which ones _should_ we keep when we're pruning them?".  And I think the answer here is to tune both the frequency and `nkeep` such that it doesn't really matter if we choose "wrong" -- i.e., if two collections start/finish at about the same time but for some reason a consumer might reasonably want either one, we should probably just keep both.  But my expectation here is that all consumers for now would probably want the same thing, which is the latest "time_started" one, and as long as "nkeep" is more than 1 then it doesn't matter which ones we keep if two overlap because there's always a newer one which is what consumers actually want.

Reply:
Ah, ok. That makes sense. I was actually thinking that we could eliminate the overlapping collections to a degree by having each nexus check that there isn't a collection currently running - or rather that one hasn't started within some bound (say collection_interval / 2) before kicking off another. With that, collections should be very close to totally ordered by time if not always so.